src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs

```diff
@@ -177,6 +177,38 @@ public override Stream GetStream(string path)
             return stream;
         }
 
+        /// <summary>
+        /// Asynchronously copies a package to the specified destination file path.
+        /// </summary>
+        /// <param name="nupkgFilePath">The destination file path.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A task that represents the asynchronous operation.
+        /// The task result (<see cref="Task{TResult}.Result" />) returns a <see cref="string" />.</returns>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="nupkgFilePath" />
+        /// is either <c>null</c> or an empty string.</exception>
+        /// <exception cref="OperationCanceledException">Thrown if <paramref name="cancellationToken" />
+        /// is cancelled.</exception>
+        public override async Task<string> CopyNupkgAsync(
+            string nupkgFilePath,
+            CancellationToken cancellationToken)
```

Comment:
@NuGet/nuget-client 
This overload was not implemented. Was it intentional? Now I can see following unit tests are start failing after I implemented this.

https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageFolderReaderTests.cs#L1073-L1083

https://github.com/NuGet/NuGet.Client/blob/936eff1f2cccbb25ba537e47a2eaaa4984a160f4/test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageArchiveReaderTests.cs#L1591-L1601

Reply:
If you do blame on the test + change, you'll see that https://github.com/NuGet/NuGet.Client/commit/870340673ba75a95407ab55ce1f46f57fe9bf25b is where this was added. 

The async copying was added for the download plugin, which is why it's the only implementation that has an async implementation. https://github.com/NuGet/NuGet.Client/blob/0b357e8339b2366f5f69b5a86a0e8d92d78f4f32/src/NuGet.Core/NuGet.Protocol/Plugins/PluginPackageReader.cs#L925-L962

So I'd say that this was ignored on purpose, because it's not a scenario we use.

Reply:
beyond that, I don't love this API, it's kind of weird in the context of the package archive reader copying its nupkg anywhere, but I won't block it based on that.

Reply:
@nkolev92 
I agree it's not really nice to have API. 
But for the sake of completeness I implemented and fixed tests. Please review again.